### ISSUE TITLE
[bottleneck issue] bottleneck issue since modelmapper version 2.1

### ISSUE THREAD

**Issue by heesuk-ahn** [modelmapper/modelmapper#520]
# [Background]

Not long ago, I had a thread bottleneck in model-mapper 2.3.1 in spring boot 2 project.

While debugging, I got to know that the issue started with version 2.1.

Starting with version 2.1, the logic for `PropertyInfoRegistry.accessorFor` has changed.

When converting from source to destination,  `PropertyInfoRegistry.accessorFor` becomes call.

In normal operation, to get accessor information, we need to get CACHE information in CuncurrentHashMap and return it immediately.

However, every time call the synchronized function repeatedly, A bottleneck has occurred.

# [Problem code]

<img width="1327" alt="issue" src="https://user-images.githubusercontent.com/42938941/74249293-4d5c0f00-4d2c-11ea-9a2d-32a83963f6a9.png">

In my case, there is `ACCESSOR CACHE` as above, but there is no `FILED_CACHE`, so I enter the section calling the loop.
 This creates a bottleneck every time the modelmapper source to destination is converted.

I've removed `FIELD_CACHE` and packaged it as a jar to see this problem more intuitively.
In this case, the bottleneck problem disappeared.

But I don't know in which case this `FIELD_CACHE` is used, so I'd like to hear from yours.

I am currently using the modelmapper version down to 2.0.0 temporarily.

**Comment by wyzssw** [comment]
+1

**Comment by jhalterman** [comment]
Happy to review a PR for this.

**Comment by jhalterman** [comment]
@heesuk-ahn @wyzssw Can you confirm that the problem is fixed for you in master?

**Comment by wyzssw** [comment]
> @heesuk-ahn @wyzssw Can you confirm that the problem is fixed for you in master?

I run junit test for 2.3.5 vs master

```
private static final ModelMapper DEFAULT_MAPPER = new ModelMapper();

 ExecutorService executorService = Executors.newFixedThreadPool(100);
        for (int i = 0; i < 20 ; i ++ ){
            executorService.submit(() -> {
                while (true) {
                    SomeBean sourceBean = new SomeBean();
                    //getter、setter
                         .....
                 
                   SomeBean destBean = new SomeBean();
                     //getter、setter
                         .....
                    DEFAULT_MAPPER.copy(sourceBean ,destBean);
                    Thread.sleep(1);
                }
            });
        }
        Thread.sleep(10000000);
```

when use 2.3.5 
many thread blocked at
```
"pool-1-thread-2" #14 prio=5 os_prio=31 tid=0x00007f8849b44800 nid=0x5903 waiting for monitor entry [0x0000700001c46000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.modelmapper.internal.PropertyInfoRegistry.accessorFor(PropertyInfoRegistry.java:100)
	- waiting to lock <0x00000006c0020540> (a java.lang.Class for org.modelmapper.internal.PropertyInfoRegistry)
	at org.modelmapper.internal.PropertyInfoRegistry.accessorFor(PropertyInfoRegistry.java:83)
```


when use master ，the blocked thread has disappear， and the mapping result is right

the problem has fixed in master, please deploy to maven central

**Comment by heesuk-ahn** [comment]
This works fine. Major changes to the modifications are as follows.

```
-    if (!ACCESSOR_CACHE.containsKey(key) || !FIELD_CACHE.containsKey(key)) {
+    if (!ACCESSOR_CACHE.containsKey(key) && !FIELD_CACHE.containsKey(key)) {
```

I also confirmed that the cache works normally through the integration test after the modification.
I think it's okay to go this way.

**Comment by jhalterman** [comment]
Fixed in 2.3.7.
